File: destack-py/destack/language/style/effect.py
from datetime import timedelta
from typing import TYPE_CHECKING, Optional

from destack.language.core import (
    Axis3,
    Enum,
    EnumType,
    NodeType,
    StructFrozen,
    StructType,
    Vector2,
    builtin_enum,
    builtin_node,
    builtin_property,
    builtin_struct,
)

from .style import Style
from .transition import Transition

if TYPE_CHECKING:
    pass


# pyright: reportIncompatibleVariableOverride=false


@builtin_enum(EnumType.EFFECT_TYPE)
class EffectType(Enum):
    """When the effect fires."""

    APPEAR = 10, "Appear", "Initial render in"
    ENTER = 11, "Enter", "Enters viewport"
    EXIT = 12, "Exit", "Leaves viewport"
    HOVER = 20, "Hover", "While hover"
    PRESS = 21, "Press", "While tap"
    DRAG = 22, "Drag", "While drag / drag"
    FOCUS = 23, "Focus", "While focus"
    LOOP = 30, "Loop", "Continuous loop"
    # SCROLL, ...


@builtin_enum(EnumType.REPEAT_TYPE)
class RepeatType(Enum):
    LOOP = 1, "Loop", "Restart from beginning"
    REVERSE = 2, "Reverse", "Yoyo back and forth"
    MIRROR = 3, "Mirror", "Mirror keyframes"


@builtin_enum(EnumType.TEXT_SPLIT_TYPE)
class TextSplitType(Enum):
    CHAR = 1, "Char", "Split by character"
    WORD = 2, "Word", "Split by word"
    LINE = 3, "Line", "Split by line"


@builtin_enum(EnumType.OFFSCREEN_BEHAVIOR)
class OffscreenBehavior(Enum):
    """What happens when the element is offscreen."""

    PLAY = 1, "Play", "Play the animation"
    PAUSE = 2, "Pause", "Pause the animation"


@builtin_struct(StructType.EFFECT, frozen=True)
class Effect(StructFrozen):
    """An effect value."""

    type: EffectType = builtin_property(30, is_repr=True)
    style: Optional["EffectStyle"] = builtin_property(41, is_repr=True)
    opacity: Optional[float] = builtin_property(50, is_repr=True)
    offset: Optional[Vector2] = builtin_property(51, is_repr=True)
    scale: Optional[float] = builtin_property(52, is_repr=True)
    rotate: Optional[Axis3] = builtin_property(53, is_repr=True)
    skew: Optional[Vector2] = builtin_property(54, is_repr=True)
    perspective: Optional[float] = builtin_property(55, is_repr=True)
    delay: Optional[timedelta] = builtin_property(56, is_repr=True)
    duration: Optional[float] = builtin_property(57, is_repr=True)
    threshold: Optional[float] = builtin_property(58, is_repr=True)
    once: Optional[bool] = builtin_property(59, is_repr=True)
    repeat: Optional[RepeatType] = builtin_property(60, is_repr=True)
    split: Optional[TextSplitType] = builtin_property(61, is_repr=True)
    offscreen: Optional[OffscreenBehavior] = builtin_property(62, is_repr=True)
    transition: Optional["Transition"] = builtin_property(70, is_repr=True)


@builtin_node(NodeType.EFFECT_STYLE)
class EffectStyle(Style):
    """An effect style."""

    type: EffectType = builtin_property(30, is_repr=True)
    opacity: Optional[float] = builtin_property(50, is_repr=True)
    offset: Optional[Vector2] = builtin_property(51, is_repr=True)
    scale: Optional[float] = builtin_property(52, is_repr=True)
    rotate: Optional[Axis3] = builtin_property(53, is_repr=True)
    skew: Optional[Vector2] = builtin_property(54, is_repr=True)
    perspective: Optional[float] = builtin_property(55, is_repr=True)
    delay: Optional[timedelta] = builtin_property(56, is_repr=True)
    duration: Optional[float] = builtin_property(57, is_repr=True)
    threshold: Optional[float] = builtin_property(58, is_repr=True)
    once: Optional[bool] = builtin_property(59, is_repr=True)
    repeat: Optional[RepeatType] = builtin_property(60, is_repr=True)
    split: Optional[TextSplitType] = builtin_property(61, is_repr=True)
    offscreen: Optional[OffscreenBehavior] = builtin_property(62, is_repr=True)
    transition: Optional["Transition"] = builtin_property(70, is_repr=True)

File: destack-rs/destack/src/imagination/animation/effect.rs
//! destack.imagination.animation.effect@2025.08.15.1

#![destack::partial(destack.imagination.animation.effect, file)]

use crate::{Axis3, Duration, Transition, Vector2};

#[destack::generated(Effect, -, block)]
/// An effect value.
pub struct Effect {
    pub r#type: EffectType,
    pub template: Option<i64>,
    pub opacity: Option<f32>,
    pub offset: Option<Vector2>,
    pub scale: Option<f32>,
    pub rotate: Option<Axis3>,
    pub skew: Option<Vector2>,
    pub perspective: Option<f32>,
    pub delay: Option<Duration>,
    pub duration: Option<f32>,
    pub threshold: Option<f32>,
    pub once: Option<bool>,
    pub repeat: Option<RepeatType>,
    pub split: Option<TextSplitType>,
    pub offscreen: Option<OffscreenBehavior>,
    pub transition: Option<Transition>,
}

#[destack::generated(RepeatType, -, block)]
/// RepeatType
pub enum RepeatType {
    /// Restart from beginning
    Loop = 1,
    /// Yoyo back and forth
    Reverse = 2,
    /// Mirror keyframes
    Mirror = 3,
}

#[destack::generated(TextSplitType, -, block)]
/// TextSplitType
pub enum TextSplitType {
    /// Split by character
    Char = 1,
    /// Split by word
    Word = 2,
    /// Split by line
    Line = 3,
}

#[destack::generated(OffscreenBehavior, -, block)]
/// What happens when the element is offscreen.
pub enum OffscreenBehavior {
    /// Play the animation
    Play = 1,
    /// Pause the animation
    Pause = 2,
}

#[destack::generated(EffectType, -, block)]
/// When the effect fires.
pub enum EffectType {
    /// Initial render in
    Appear = 10,
    /// Enters viewport
    Enter = 11,
    /// Leaves viewport
    Exit = 12,
    /// While hover
    Hover = 20,
    /// While tap
    Press = 21,
    /// While drag / drag
    Drag = 22,
    /// While focus
    Focus = 23,
    /// Continuous loop
    Loop = 30,
}
